### DISCUSSION TITLE
Recommender System

### DISCUSSION THREAD

**Discussion by RetrosynthesisAI** [pyg-team/pytorch_geometric#4022] [Q&A]
I am interested in using the HetNet Link Prediction code to develop a recommender system, that prints out batches of recommendations for each input dataset. Here is your HetNet Link Prediction code:

https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/hetero_link_pred.py

Do you have any recommendations on how to go about this?

**Comment by rusty1s** [comment]
Can you be a bit more specific about what you try to achieve? Do you want to train your recommender system in a mini-batch fashion or simply want to evaluate it in batches?

**Comment by RetrosynthesisAI** [comment]
Sure, I am interested in creating a recommendation system (full batch). I would like to recommend safer medications to users based on a safety rating and meta data with a graph network. I would like to start working towards this aim with the Het Link Prediction code that you offer in your github. But don't know how to start.

Abstract: 
https://eejournal.ktu.lt/index.php/elt/article/download/24828/12890

**Reply by rusty1s** [reply]
I would start by trying to map your dataset to the `HeteroData` object, e.g., encoding features into feature matrices, creating edge indices, etc. Given that, I think there might be only minor modifications to the example needed to let it execute on your specific data. Please feel free to share more data information or whether you need more help to map it to `HeteroData`.

**Reply by RetrosynthesisAI** [reply]
Thanks. Also I was wondering if you have a function for using the created HetNet Link Prediction model with a batch dataset.
Something like:

```
prediction = model.predict(batch_dataset, gpu=4)
```

Where the output is a dataframe (prediction results)
```
|    Predicted Categorical Label |   Probability   |  Comment about Prediction | 
------------------------------------------------------------------------------
|    Safe      |  0.99 | any string |
|    Not Safe  |  0.85 | any string |
|    Safe      |  0.55 | any string |
```

**Reply by rusty1s** [reply]
No, something like that is currently not supported. We currently recommend to use PyTorch Lightning for any multi-GPU scenarios. You can find some examples [here](https://github.com/pyg-team/pytorch_geometric/tree/master/examples/pytorch_lightning).
